### PR TITLE
[Perf][TTS] glm_tts: hoist loop-invariant text embed, RoPE, mask and CFG batch out of Euler loop

### PR DESCRIPTION
**Pull request by JuanPZuluaga**
## Purpose

GLM-TTS flow-matching sampling runs a 10-step Euler/CFM loop where each step recomputed several inputs that never change across steps: the text embedding (token embedding + ConvNeXt stack), the RoPE table, the block-causal attention mask, and the CFG double-batch conditioning tensors (`cat`/`expand` of condition, speech tokens, padding mask, speaker embedding). This PR computes them once per `_do_sample` call and passes them into the DiT,
removing ~(n_timesteps-1) redundant evaluations and the per-step CFG tensor-construction launches. The eager path is launch-bound at low concurrency, so this reduces per-request host overhead; it also applies to the streaming (block-causal) path, which cannot use the CUDA-graph fast path.

the changes: 

- `GLMTTSDiT.forward` accepts optional `text_embed` and `rope`; when provided it skips the   internal recompute (defaults preserve existing behavior for all other callers, including   the CUDA-graph wrapper).
- `GLMTTSDiTForGeneration._do_sample` hoists out of the Euler loop:
  - text embedding (`text_emb_layer(...)` + model-dtype cast) for both CFG and non-CFG branches
  - RoPE table (`rotary_embed.forward_from_seq_len(seq_len)`)
  - block-causal attention mask (passed as `attn_mask`, replacing per-step
    `is_causal`/`block_pattern` rebuild)
  - CFG double-batch tensors (`condition_b`, `text_b`, `padding_b`, `spkr_b`, `text_lens_b`)

## Numbers

Measured wall-clock, branch vs base (origin/main), latest code (vllm 0.25.0), same server/venv, production async-chunk deploy yaml (2 stages, one RTX 6000 Ada). Four independent runs; the largest run is 40 requests per variant at c=8 and 80 at c=16.

| concurrency | throughput (audio-s/s) | e2e p50 | e2e p99 |
|---|---|---|---|
| c=1 | ~flat | ~flat | — |
| c=8 | **+4.7% .. +11.2%** faster (4/4 runs positive) | **−2.8% .. −9.0%** | up to −18% |
| c=16 | ~flat (−1.8% .. +6.0%) | −2.1% .. −6.5% | — |

The win concentrates at moderate concurrency (c≈8), where per-request host-side setup in the Euler loop competes with GPU work; at c=16 the GPU is saturated and the hoist is neutral. Byte-identical output.

## Test Plan

**vLLM Version:**

**vLLM-Omni Commit:**

## Test Result

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by linyueqian** [comment]
@BeatSeat do you have bandwidth to test on this?

**Comment by BeatSeat** [comment]
Hi, why this cannot use the CUDA Graph path?

File: vllm_omni/model_executor/models/glm_tts/glm_tts_dit.py [hunk 1]
```diff
@@ -610,13 +615,15 @@ def forward(
             time_emb = torch.cat([time_emb, spkr_emb], dim=-1)
 
         # Text embedding - ensure cast to model dtype
-        text_embed = self.text_emb_layer(text, seq_len, text_lens=text_lens).to(model_dtype)
+        if text_embed is None:
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/glm_tts/glm_tts_dit.py:618]: I like the escape hatch, and I checked that `text_emb_layer` only depends on text/seq_len/text_lens so precomputing it is genuinely safe. One thing I keep coming back to though: once a caller supplies `text_embed` (and `rope`), `text` is silently ignored and `rope` is trusted to match `seq_len = middle_point.shape[1]`. If a precomputed rope/embed ever gets built for a different length than `middle_point`, the transformer just applies mismatched rotary and produces quiet garbage. Since the whole hoist rests on that assumption, I'd add a cheap guard here, e.g. `assert text_embed is None or text_embed.shape[1] == seq_len`. 🤔

File: vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py [hunk 2]
```diff
@@ -857,6 +857,36 @@ def _do_sample(
         if self.t_scheduler == "cosine":
             t_span = 1 - torch.cos(t_span * 0.5 * torch.pi)
 
+        # Hoist loop-invariant DiT inputs (text embedding, RoPE table,
+        # block-causal mask, CFG double-batch conditioning) out of the Euler
+        # loop; they depend only on the conditioning and seq_len, not on the
+        # per-step state x or the timestep.
+        seq_len = int(mel_cond.shape[1])
+        dit_dtype = self.dit.proj_out.weight.dtype
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py:867]: Took me a sec but I confirmed `dit_dtype` here equals what `forward()` uses internally (both are `proj_out.weight.dtype`), so the hoisted `.to(dit_dtype)` casts match the old in-`forward` cast exactly. It is a duplicated source of truth now though - if `forward`'s `model_dtype` derivation ever changes, these hoisted casts would silently diverge. Might be worth a short comment tying it back to `forward`'s `model_dtype` so nobody unpicks one without the other.

File: vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py [hunk 3]
```diff
@@ -857,6 +857,36 @@ def _do_sample(
...
+        attn_mask = (
+            self.dit._create_block_causal_mask(block_pattern, seq_len, device, dit_dtype)
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py:871]: One thing I want to double-check: the old path built this mask inside `forward` on `middle_point.device`, whereas here you pass `device` (= `speech_tokens.device`). I'm fairly sure they're identical in this call path, but is that actually guaranteed? If speech tokens and the sampled `x` could ever land on different devices this would be the one spot that breaks. 👍

**Comment by JuanPZuluaga** [comment]
Thanks @BeatSeat: the streaming/block-causal path always sets `block_pattern`, which bypasses the DiT CUDA graph, so it runs the eager Euler loop. This hoist speeds up exactly that case, the graphed (non-streaming) runs hit the early-return and are untouched.

File: vllm_omni/model_executor/models/glm_tts/glm_tts_dit.py (hunk 1 above)
Comment (JuanPZuluaga) [reply vllm_omni/model_executor/models/glm_tts/glm_tts_dit.py:618]: Added the guard

File: vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py (hunk 2 above)
Comment (JuanPZuluaga) [reply vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py:867]: Added a comment tying `dit_dtype` back to `forward()`'s `model_dtype` (both `proj_out.weight.dtype`) so they can't drift.

File: vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py (hunk 3 above)
Comment (JuanPZuluaga) [reply vllm_omni/model_executor/models/glm_tts/glm_tts_dit_wrapper.py:871]: i confirmed this, the call path the sampled `x` lives on `speech_tokens.device`, so the mask device matches the old in-`forward` one.

**Comment by JuanPZuluaga** [comment]
@lishunyang12 I updated the code with your comments. Thanks!

**Comment by linyueqian** [comment]
@vllm-omni-review-bot

**Comment by linyueqian** [comment]
A/B on an L20X box (vLLM 0.26.0, `glm_tts.yaml`, GLM-TTS, voice-clone request with ref_audio+ref_text). Baseline is `pr5068^2` = `8e44e780`, so the legs differ by exactly this PR. Two independent server instances per leg, warm requests only (first request after load discarded as cold):

| leg | warm samples (s) | median |
|---|---|---|
| baseline `8e44e780` | 2.33, 2.39, 2.42, 2.42, 2.35, 2.38 | **2.39** |
| PR head `e10a38a1` | 2.08, 2.10, 2.09, 2.05, 2.12, 2.11 | **2.10** |

**~12% faster end-to-end, and the ranges don't overlap** (PR max 2.12 < baseline min 2.33) across separate server starts, so this isn't load drift. Nice result — the Euler-loop hoisting shows up clearly even on a single-request wall-clock probe.

Correctness: output size identical (460844 B) in every run, no failures in 16 requests. Digests are mostly identical across legs too — requests 1/2/3 produced byte-identical audio on both sides; only the cold request and one later one differed, which is what I'd expect from sampling-state drift rather than from the math changing. So the hoisting looks numerically faithful, not just close.

Not covered by this: I only exercised the default single-request path, so nothing here speaks to the CFG-batch change under concurrency. Happy to re-run with a concurrency sweep if that's where you expect the bigger win.

**Review by linyueqian — APPROVED** [review]
Verified by A/B on an L20X box against `pr5068^2`: ~12% faster end-to-end (median 2.10s vs 2.39s), non-overlapping ranges across two independent server instances per leg, and requests 1-3 produced byte-identical audio on both sides — so the Euler-loop hoisting is numerically faithful, not just close. No failures in 16 requests. Details in my earlier comment.

**Comment by linyueqian** [comment]
#5067 and #5175 merged; this one is held up by a single docs warning. `buildkite/vllm-omni` is green, and the AMD leg is red on all three (it was red on both PRs that merged), so readthedocs is the only thing actually blocking:

```
python -m mkdocs build --clean ... -> exit 1
WARNING - griffe: vllm_omni/model_executor/models/glm_tts/glm_tts_dit.py:595:
          No type or annotation for parameter 'rope'
Aborted with 1 warnings in strict mode!
```

mkdocs runs strict, so one griffe warning fails the build. The new `rope=None` parameters (lines 335, 409, 579) have no annotation while the docstring documents them, which is the combination griffe complains about. Given the unpack at line 346 (`freqs, xpos_scale = rope`), something like:

```python
rope: tuple[torch.Tensor, torch.Tensor] | None = None,
```

on the three signatures should clear it.

Happy to push that myself if you'd rather not context-switch — maintainer edits are enabled on the branch, just say the word. Otherwise it's yours; the approval and auto-merge are still armed, so it'll go in on its own once the docs build passes.

**Comment by linyueqian** [comment]
@JuanPZuluaga could you push the annotation fix when you get a chance? Adding the type to the three `rope=None` signatures (lines 335, 409, 579) is all that's between this and merge — approval and auto-merge are armed, so it'll go in by itself once readthedocs is green. This is the ~12% one, so it'd be nice to land.

**Comment by JuanPZuluaga** [comment]
@linyueqian thanks for all the comments. I just pushed some changes, could you check again if all is correct? :)
